clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreBufferService.java

```diff
@@ -516,6 +546,66 @@ public boolean equals(Object o) {
     }
   }
 
+  private static class CommandQueueNode extends QueueNode {
+    enum CommandType {
+      // only supports SYNC_OFFSET command today.
+      SYNC_OFFSET
+    }
+
+    private final LockAssistedCompletableFuture<Void> cmdExecutedFuture;
+
+    private final CommandType commandType;
+
+    public CommandQueueNode(
+        CommandType commandType,
+        PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
+        StoreIngestionTask ingestionTask) {
+      super(consumerRecord, ingestionTask, StringUtils.EMPTY, 0);
+      this.commandType = commandType;
+      this.cmdExecutedFuture = new LockAssistedCompletableFuture<>(this);
+    }
+
+    public CompletableFuture<Void> getCmdExecutedFuture() {
+      return cmdExecutedFuture;
+    }
+
+    public CommandType getCommandType() {
+      return commandType;
+    }
+
+    /*
+     * This method is used to execute the command synchronously. We use a customized LockAssistedCompletableFuture to
+     * ensure that once it is in execution, it cannot be cancelled.
+     */
+    public void executeSync(Runnable runnable) {
+      synchronized (cmdExecutedFuture.getLock()) {
+        if (cmdExecutedFuture.isDone() || cmdExecutedFuture.isCancelled()) {
+          LOGGER.warn(
+              "Command {} for {} in drainer queue is already done or cancelled",
+              commandType,
+              getConsumerRecord().getTopicPartition());
+          return;
+        }
+        runnable.run();
```

Comment:
Should we wrap it up by `try/catch` here with proper logging?
```
try {
    runnable.run();
    cmdExecutedFuture.complete(null);
} catch (Exception e) {
    cmdExecutedFuture.completeExceptionally(e);
}

Reply:
agree and it looks in a better shape.